# Web Scraping of Rating Prediction

In [1]:
#Importing Necessary Libraries
import pandas as pd
import selenium
import requests
import time
from selenium import webdriver
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
import warnings
warnings.filterwarnings('ignore')


C:\Users\Aamina\anaconda3\lib\site-packages\requests\__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.8) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)


# Scraping Review data from Amazon

In [2]:
#Opening the Edge browser
driver=webdriver.Edge('msedgedriver.exe')

In [3]:
url='https://www.amazon.in/'
driver.get(url)

In [4]:
driver.maximize_window()

In [122]:
url = "https://www.amazon.in/"
driver.get(url)
time.sleep(3)

In [123]:
driver.maximize_window()

In [119]:
search_items = ['laptops','phones','headphones','smart watches', 'Professional Cameras', 'Printers','Home theater']

In [124]:
title = []
review_text = []
ratings = []

In [ ]:
search_items = ['phones','headphones','smart watches', 'Professional Cameras', 'Printers','Home theater']
for i in search_items:
    srchbar = driver.find_element_by_id("twotabsearchtextbox")
    srchbar.clear()
    srchbar.send_keys(i)
    driver.find_element_by_id("nav-search-submit-button").click()  #clicking on search button
    time.sleep(1)
    item_url = []
    for i in driver.find_elements_by_xpath("//h2[@class='a-size-mini a-spacing-none a-color-base s-line-clamp-2']/a"):
        item_url.append(i.get_attribute('href'))
    for i in item_url:
        driver.get(i)
        time.sleep(1)
        for _ in range(2):
            driver.execute_script("window.scrollBy(0,6000)")
            time.sleep(1)
        #click on to see all reviews
        try:
            driver.find_element_by_xpath("//a[@class='a-link-emphasis a-text-bold']").click()
        except:
            pass
        
        url_ = []
        try:
            two_str = driver.find_element_by_xpath("//table[@id='histogramTable']/tbody/tr[4]/td[2]/a")
            url_.append(two_str.get_attribute('href'))
        except: pass
        try:
            three_str = driver.find_element_by_xpath("//table[@id='histogramTable']/tbody/tr[3]/td[2]/a")
            url_.append(three_str.get_attribute('href'))
        except: pass
        try:
            one_str = driver.find_element_by_xpath("//table[@id='histogramTable']/tbody/tr[5]/td[2]/a")
            url_.append(one_str.get_attribute('href'))
        except: pass
        try:
            fiv_str = driver.find_element_by_xpath("//table[@id='histogramTable']/tbody/tr[1]/td[2]/a")
            url_.append(fiv_str.get_attribute('href'))
        except: pass
        try:
            four_str = driver.find_element_by_xpath("//table[@id='histogramTable']/tbody/tr[2]/td[2]/a")
            url_.append(four_str.get_attribute('href'))
        except: pass
        
        for i in url_:
            driver.get(i)
            time.sleep(1)
            for i in range(10):
                ids = driver.find_elements_by_xpath("//div[@class='a-section a-spacing-none review-views celwidget']/div")
                comment_ids = []
                for i in ids:
                    comment_ids.append(i.get_attribute('id'))

                for x in comment_ids:
                    #Extract  ids from each user on a page
                    #Scarping the Review title
                    try:
                        review_title = driver.find_element_by_xpath('//*[@id="' + x +'"]/div/div/div[2]/a[2]/span[1]')
                        title.append(review_title.get_attribute("innerHTML"))
                    except NoSuchElementException:
                        title.append('')
                        
                    #Scraping the Review Ratings
                    try:
                        rating = driver.find_element_by_xpath('//*[@id="' + x +'"]/div/div/div[2]/a[1]/i/span[1]')
                        ratings.append(rating.get_attribute("innerHTML"))
                    except NoSuchElementException:
                        ratings.append('')
                   
                    #Scraping the Review Text
                    try:
                        text = driver.find_element_by_xpath('//*[@id="' + x +'"]/div/div/div[4]/span/span')
                        review_text.append(text.get_attribute("innerHTML"))
                    except NoSuchElementException:
                        review_text.append('')
                try:
                    driver.find_element_by_xpath("//ul[@class='a-pagination']/li[2]/a").click()
                    time.sleep(3)
                except : break

In [130]:
df1=pd.DataFrame()
df1['Review Title']=title
df1['Review Text']=review_text
df1['Ratings']=ratings
df1

Review Title  \
0             Don't even have Charger with this product   
1                   Handset made with old used material   
2                               Camera quality not good   
3     Without charger it's not worth for this much a...   
4     Battery consumption is very high.. and it is g...   
...                                                 ...   
1721  Overall a good product fingerprint was lacking...   
1722                                               Good   
1723                            Good for oldage people,   
1724                                                The   
1725                                                      

                                            Review Text             Ratings  
0     &nbsp;Better to provide all basic accessories ...  2.0 out of 5 stars  
1     Not satisfactory!<br><br>1. It seem side panel...  2.0 out of 5 stars  
2     Camera quality is not good, it is not worth fo...  2.0 out of 5 stars  
3                               Charger should included  2.0 out of 5 stars  
4     I think It could be more better.. doesn't meet...  2.0 out of 5 stars  
...                                                 ...                 ...  
1721  A good budget smartphone for all basic needs a...  4.0 out of 5 stars  
1722                                               Good  4.0 out of 5 stars  
1723  There is no finger print, battery good, displa...  4.0 out of 5 stars  
1724  Mobile bahut achcha hai battery backup bhi bad...  4.0 out of 5 stars  
1725                                                                         

[1726 rows x 3 columns]

In [131]:
df1.to_csv('phone_reviews.csv')

In [8]:
df2=pd.DataFrame()
df2['Review Title']=title
df2['Review Text']=review_text
df2['Ratings']=ratings
df2

Review Title  \
0      Can't use these headphones for more than an hour   
1     below average sound quality and bass level is ...   
2                   Disappointed with Bluetooth headset   
3     This headphone was a let down. The worst is th...   
4                                  Bad quality product.   
...                                                 ...   
8928                                     Decent product   
8929                                 Noce for the price   
8930                         Product performance is gud   
8931                                               Good   
8932                                                      

                                            Review Text             Ratings  
0     I bought this headphone as a replacement for m...  2.0 out of 5 stars  
1     connectivity - good<br>comfort level - good<br...  2.0 out of 5 stars  
2     I was expected a lot from this Bluetooth heads...  2.0 out of 5 stars  
3     Heres my initial experience...(as a profession...  2.0 out of 5 stars  
4     Bought this product on the launch day, with lo...  2.0 out of 5 stars  
...                                                 ...                 ...  
8928                                    Value for money  4.0 out of 5 stars  
8929  Build quality os okay but not great, sound is ...  4.0 out of 5 stars  
8930  The product performance is gud and delivered i...  4.0 out of 5 stars  
8931                                          Very well  4.0 out of 5 stars  
8932                                                                         

[8933 rows x 3 columns]

In [9]:
df2.to_csv('head_phones.csv')

In [10]:
dfa1=pd.read_csv('Laptop_Review.csv')
dfa1.drop(columns=['Unnamed: 0'],inplace=True)

In [11]:
dfa2=pd.read_csv('phone_reviews.csv')
dfa2.drop(columns=['Unnamed: 0'],inplace=True)

In [14]:
dfa3=pd.read_csv('head_phones.csv')
dfa3.drop(columns=['Unnamed: 0'],inplace=True)

In [15]:
amazon=pd.concat([dfa1,dfa2,dfa3],axis=0)
amazon

Review Title  \
0                              Suitable for School kids   
1     Misrepresentation on MS Office 2019 license - ...   
2                            The sold me renewed laptop   
3        Amazon dupes with specification/ battery sucks   
4               Low Quality and Low Battery performance   
...                                                 ...   
8928                                     Decent product   
8929                                 Noce for the price   
8930                         Product performance is gud   
8931                                               Good   
8932                                                NaN   

                                            Review Text             Ratings  
0     If you are a College student or a professional...  2.0 out of 5 stars  
1     Update after one month usage - MS Office 2019 ...  2.0 out of 5 stars  
2     It’s look like renewed laptop because laptop c...  2.0 out of 5 stars  
3     &nbsp;I had seen the specifications and bought...  2.0 out of 5 stars  
4     Build Quality was Low, No match for the price....  2.0 out of 5 stars  
...                                                 ...                 ...  
8928                                    Value for money  4.0 out of 5 stars  
8929  Build quality os okay but not great, sound is ...  4.0 out of 5 stars  
8930  The product performance is gud and delivered i...  4.0 out of 5 stars  
8931                                          Very well  4.0 out of 5 stars  
8932                                                NaN                 NaN  

[11907 rows x 3 columns]

In [16]:
amazon.to_csv('Amazon_Review.csv')

# Scraping Review data from Flipkart

In [10]:
url = "https://www.flipkart.com/"
driver.get(url)
time.sleep(2)
#clicking on cancel buttopn 
driver.find_element_by_xpath("//div[@class='_2QfC02']/button").click()

In [26]:
search_items = ['Phones','smart watches','Monitors','printers','headphones', 'Professional Cameras']

In [27]:
title = []
review_text = []
ratings = []
def scrap():    
        for i in driver.find_elements_by_xpath("//div[@class='t-ZTKy']"):
            review_text.append(i.text)
        for i in driver.find_elements_by_xpath("//p[@class='_2-N8zT']"):
            title.append(i.text)
        for i in driver.find_elements_by_xpath("//p[@class='_2-N8zT']/../div"):
            ratings.append(i.text)
        return

In [28]:

urls=[]
for i in search_items:
    #Find search bar
    srchBar = driver.find_element_by_xpath("//div[@class='_3OO5Xc']/input")
    srchBar.clear()
    srchBar.send_keys(i)
    #clicking on search button
    driver.find_element_by_xpath("//button[@class='L0Z3Pu']").click()
    time.sleep(3)
    page = []
    for i in driver.find_elements_by_xpath("//nav[@class='yFHi8N']/a"):
        page.append(i.get_attribute('href'))
    for i in page[0:4]:
        driver.get(i)
        time.sleep(3)
        items = driver.find_elements_by_xpath("//a[@class='_1fQZEK']")
        for i in items:
            urls.append(i.get_attribute('href'))

In [30]:
for i in urls:
    driver.get(i)
    time.sleep(2)
    for _ in range(2):
        driver.execute_script("window.scrollBy(0,6000)")
        time.sleep(1)
        #clicking on all reviews
    try:
        btn=driver.find_element_by_xpath("//div[@class='_2c2kV-']/following::a")
        lnk = btn.get_attribute('href')
        driver.get(lnk)
        time.sleep(1)
    except NoSuchElementException:
        pass

    scrap()        
    try:
        n_page=driver.find_elements_by_xpath("//nav[@class='yFHi8N']/a")
        np=[]
        for i in n_page:
            np.append(i.get_attribute('href'))
        for i in np[0:18]:
            driver.get(i)
            time.sleep(1)
            scrap()
    except: continue

In [24]:
dff=pd.DataFrame()
dff['Review Title']=title
dff['Review Text']=review_text
dff['Ratings']=ratings
dff

Review Title                                        Review Text  \
0          Good choice  Great Laptop, performance is too good,\nMetal ...   
1            Wonderful  Laptop is superb both in performance and looks...   
2     Perfect product!  Good Laptop. Eclipse are getting loaded very f...   
3                 Good  Battery not good as per they say\n\nAnd warran...   
4            Just wow!  Its ultimate in every way. Totally pro and bus...   
...                ...                                                ...   
3865         Must buy!  My 17 days experience this asus laptop is amaz...   
3866       Pretty good  A Good looking, easy to carry move mediocre pe...   
3867          Terrific  Awesome choice for Normal usage...\nGood displ...   
3868         Fabulous!                                        nice design   
3869          Terrific  Used for a week till now.\nProduct is awesome,...   

     Ratings  
0          4  
1          5  
2          5  
3          3  
4          5  
...      ...  
3865       5  
3866       4  
3867       5  
3868       5  
3869       5  

[3870 rows x 3 columns]

In [25]:
dff.to_csv('flipkart_laptop.csv')

In [32]:
dff=pd.DataFrame()
dff['Review Title']=title
dff['Review Text']=review_text
dff['Ratings']=ratings
dff

Review Title  \
0                 Just wow!   
1         Terrific purchase   
2             Great product   
3            Classy product   
4      Good quality product   
...                     ...   
19392     Terrific purchase   
19393       Worth the money   
19394             Excellent   
19395        Simply awesome   
19396               Awesome   

                                             Review Text Ratings  
0      In this price.....this phone is fantastic guys...       5  
1      First impression is very good . Value for mone...       5  
2      Thank you for Flipkart again for good conditio...       5  
3      This is a very premium looking mobile with lar...       5  
4      Good value for money product. Camera I not tha...       4  
...                                                  ...     ...  
19392  Excellent monitor got it at great price of 720...       5  
19393                  Good, but rate high when compared       4  
19394                                            Super Q       5  
19395                                      Super se uper       5  
19396                      Worthy Offer, monitor is good       5  

[19397 rows x 3 columns]

In [33]:
dff.to_csv('Flipkart_reviews.csv')

In [18]:
df1=pd.read_csv('flipkart_laptop.csv')


In [3]:
df1.drop(columns='Unnamed: 0',inplace=True)

In [17]:
df2=pd.read_csv('Flipkart_reviews.csv')


In [6]:
df2.drop(columns='Unnamed: 0',inplace=True)

In [7]:
flipkart=pd.concat([df1,df2],axis=0)
flipkart

Review Title                                        Review Text  \
0            Good choice  Great Laptop, performance is too good,\nMetal ...   
1              Wonderful  Laptop is superb both in performance and looks...   
2       Perfect product!  Good Laptop. Eclipse are getting loaded very f...   
3                   Good  Battery not good as per they say\n\nAnd warran...   
4              Just wow!  Its ultimate in every way. Totally pro and bus...   
...                  ...                                                ...   
19392  Terrific purchase  Excellent monitor got it at great price of 720...   
19393    Worth the money                  Good, but rate high when compared   
19394          Excellent                                            Super Q   
19395     Simply awesome                                      Super se uper   
19396            Awesome                      Worthy Offer, monitor is good   

       Ratings  
0            4  
1            5  
2            5  
3            3  
4            5  
...        ...  
19392        5  
19393        4  
19394        5  
19395        5  
19396        5  

[23267 rows x 3 columns]

In have scraped 23267 rows and 3 columns reviews data from Flipkart

In [12]:
flipkart.to_csv('Flipkart_Review.csv')

In [19]:
total_reviews=pd.concat([amazon,flipkart],axis=0)
total_reviews

Review Title  \
0                               Suitable for School kids   
1      Misrepresentation on MS Office 2019 license - ...   
2                             The sold me renewed laptop   
3         Amazon dupes with specification/ battery sucks   
4                Low Quality and Low Battery performance   
...                                                  ...   
19392                                  Terrific purchase   
19393                                    Worth the money   
19394                                          Excellent   
19395                                     Simply awesome   
19396                                            Awesome   

                                             Review Text             Ratings  
0      If you are a College student or a professional...  2.0 out of 5 stars  
1      Update after one month usage - MS Office 2019 ...  2.0 out of 5 stars  
2      It’s look like renewed laptop because laptop c...  2.0 out of 5 stars  
3      &nbsp;I had seen the specifications and bought...  2.0 out of 5 stars  
4      Build Quality was Low, No match for the price....  2.0 out of 5 stars  
...                                                  ...                 ...  
19392  Excellent monitor got it at great price of 720...                   5  
19393                  Good, but rate high when compared                   4  
19394                                            Super Q                   5  
19395                                      Super se uper                   5  
19396                      Worthy Offer, monitor is good                   5  

[35174 rows x 3 columns]

In [20]:
total_reviews.to_csv('Amazon_Flipkart_Reviews.csv')

Hence successfully scraped 35174 reviews from Flipkart and Amazon